### PR TITLE
Addition of positioning tests

### PR DESCRIPTION
**Pull request by nirvn**
This PR introduces brand new positioning testing in our growing test suite. The testing relies on a newly-introduced server that streams NMEA strings through a tcp or udp socket. This builds on the work done to introduce brand new positioning TCP and UDP receivers in QField.

The tests cover the following functionality:
- Basic positioning information
- Ellispoidal height
- Delta Z (i.e. antenna height)
- Vertical grid shift

The tests are run within a QML environment to mimic QField as closely as possible. This provides us with a virtual 'armor' that will greatly reduce the risks of regressions slipping into this crucial part of QField.

@m-kuhn , I'm pretty happy about this overall. I'd love your thoughts on this, particularly on the way the nmea server is launched here.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for fd3502da3afada921c84596d84b8861f5743b415: [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3771-f1852f5057617d0c3605788ddffe0bd7d9183a6e-arm64-android.apk)
<details>
<summary>Other architectures</summary>

- [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3771-f1852f5057617d0c3605788ddffe0bd7d9183a6e-arm-android.apk)
- [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3771-f1852f5057617d0c3605788ddffe0bd7d9183a6e-x64-android.apk)
- [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3771-f1852f5057617d0c3605788ddffe0bd7d9183a6e-x86-android.apk)
- [**all access arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3771-f1852f5057617d0c3605788ddffe0bd7d9183a6e-arm64-android.apk)
- [all access arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3771-f1852f5057617d0c3605788ddffe0bd7d9183a6e-arm-android.apk)
- [all access x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3771-f1852f5057617d0c3605788ddffe0bd7d9183a6e-x64-android.apk)
- [all access x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3771-f1852f5057617d0c3605788ddffe0bd7d9183a6e-x86-android.apk)
</details>

**Comment by m-kuhn** [comment]
Glad to see this land!!

> particularly on the way the nmea server is launched here

That looks good to me 👍 

Other than that a larger review would be good. Not sure when I can do that, if someone else wants to go for this (@domi4484 e.g.?) I'm more than happy.

**Comment by nirvn** [comment]
@m-kuhn , I'm very glad too. This will dramatically reduces a crucial set of QField functionalities from regressing.

**Comment by nirvn** [comment]
@m-kuhn , tests covering delta Z (antenna height) as well as vertical grid shift added. This was a *perfect* opportunity to clean up vertical grid handling in our code (I'm the one who did the original implementation, and I hate myself for not having gone down the road of a property to begin with).

This feels really solid now.

File: src/core/positioning/nmeagnssreceiver.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  mNmeaConnection = std::make_unique<QgsNmeaConnection>( ioDevice );
+
+  //QgsGpsConnection state changed (received location string)
+  connect( mNmeaConnection.get(), &QgsGpsConnection::stateChanged, this, &NmeaGnssReceiver::stateChanged );
+}
+
+void NmeaGnssReceiver::stateChanged( const QgsGpsInformation &info )
+{
+  if ( mLastGnssPositionValid && std::isnan( info.latitude )               // we already sent a valid position
+       || info.utcDateTime == mLastGnssPositionInformation.utcDateTime() ) // we group updates by timestamp, if the last block is not finished, return
```
Comment (m-kuhn) [inline src/core/positioning/nmeagnssreceiver.cpp:38]: Is this outdated already?

File: test/qml/tst_positioning.qml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        // we therefore simply check whether we are int the 200s value range, which indicates ellispodal elevation is
+        // being returned
+        compare(Math.floor(positioning.positionInformation.elevation / 100), 2)
+    }
+
+    function test_deltaZ() {
+        coordinateTransformer.deltaZ = -100
+        coordinateTransformer.verticalGrid = ''
+        // wait a few seconds so positioning can catch some NMEA strings
+        wait(5000)
```
Comment (m-kuhn) [inline test/qml/tst_positioning.qml:43]: That's a lot of waits in all these tests... Can we reduce this to <1s, e.g. by increasing the server rate?

File: src/core/positioning/nmeagnssreceiver.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/positioning/nmeagnssreceiver.cpp:38]: Yes, please review the TCP/UDP receiver PR for thess parts, once merged I'll rebase this PR which will then only contain tests stuff.

File: test/qml/tst_positioning.qml (hunk 2 above)
Comment (nirvn) [reply test/qml/tst_positioning.qml:43]: @m-kuhn , I shall test and adjust :)

File: test/nmea_server/nmeaserver.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    "-t",
+    "--type",
+    help="Server type to operate on",
+    choices=["tcp", "udp"],
+    default="tcp",
+    type=str,
+)
+parser.add_argument("-p", "--port", help="Port to operate on", default=1958, type=int)
+parser.add_argument(
+    "-i", "--interval", help="Pause between sending sentences", default=0.1, type=float
```
Comment (nirvn) [inline test/nmea_server/nmeaserver.py:32]:
```suggestion
    "-i", "--interval", help="Pause between sending sentences", default=0.05, type=float
```

File: test/qml/tst_positioning.qml (hunk 2 above)
Comment (nirvn) [reply test/qml/tst_positioning.qml]: @m-kuhn , double server rate and decreased wait by half for each test. Tests were failing (either osx or both osx and linux at times) using lower values.

File: src/core/positioning/nmeagnssreceiver.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/positioning/nmeagnssreceiver.cpp]: Done.

File: test/nmea_server/nmeaserver.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    "-t",
+    "--type",
+    help="Server type to operate on",
+    choices=["tcp", "udp"],
+    default="tcp",
+    type=str,
+)
+parser.add_argument("-p", "--port", help="Port to operate on", default=1958, type=int)
+parser.add_argument(
+    "-i", "--interval", help="Pause between sending sentences", default=0.05, type=float
```
Comment (m-kuhn) [inline test/nmea_server/nmeaserver.py:32]:
```suggestion
    "-i", "--interval", help="Pause between sending sentences in seconds", default=0.05, type=float
```

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**